File: src/main/java/bot/inker/ankh/core/item/AbstractAnkhItem.java
package bot.inker.ankh.core.item;

import bot.inker.ankh.core.api.item.AnkhItem;
import bot.inker.ankh.core.libs.nbtapi.NBTItem;
import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import javax.annotation.Nonnull;
import java.util.List;

public abstract class AbstractAnkhItem implements AnkhItem {
  public abstract @Nonnull Material material();

  public abstract @Nonnull Component itemName();

  public abstract @Nonnull List<Component> lores();

  public ItemStack createItem() {
    ItemStack itemStack = new ItemStack(Material.CLOCK);
    updateItem(itemStack);
    return itemStack;
  }

  @Override
  public final void updateItem(ItemStack item) {
    item.setType(material());
    NBTItem nbtItem = new NBTItem(item);
    if (!key().asString().equals(nbtItem.getString(ITEM_ID_TAG))) {
      nbtItem.setString(ITEM_ID_TAG, key().asString());
    }
    onUpdateItemNbt(nbtItem);
    nbtItem.applyNBT(item);
    item.editMeta(meta -> {
      meta.displayName(itemName());
      meta.lore(lores());
      onUpdateItemMeta(meta);
    });

    onUpdateItem(item);
  }

  protected void onUpdateItem(ItemStack item) {
    //
  }

  protected void onUpdateItemNbt(NBTItem nbtItem) {
    //
  }

  protected void onUpdateItemMeta(ItemMeta itemMeta) {
    //
  }
}

File: src/main/kotlin/bot/inker/ankh/core/item/AbstractAnkhItem.kt
package bot.inker.ankh.core.item

import bot.inker.ankh.core.api.item.AnkhItem
import bot.inker.ankh.core.libs.nbtapi.NBTItem
import net.kyori.adventure.text.Component
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta

abstract class AbstractAnkhItem : AnkhItem {
  abstract val itemId: NamespacedKey
  abstract val material: Material
  abstract val itemName: Component
  abstract val lores: List<Component>

  override fun getKey(): NamespacedKey = itemId

  fun createItem(): ItemStack {
    return ItemStack(Material.CLOCK).also(this::updateItem)
  }

  final override fun updateItem(item: ItemStack) {
    item.type = material

    NBTItem(item).apply {
      if (itemId.asString() != getString("ankh-core:item-id")) {
        setString("ankh-core:item-id", itemId.asString())
      }
      onUpdateItemNbt(this)
    }.applyNBT(item)

    item.editMeta { meta ->
      meta.displayName(itemName)
      meta.lore(lores)
      onUpdateItemMeta(meta)
    }

    onUpdateItem(item)
  }

  open fun onUpdateItem(item: ItemStack) {
    //
  }

  open fun onUpdateItemNbt(nbtItem: NBTItem) {
    //
  }

  open fun onUpdateItemMeta(itemMeta: ItemMeta) {
    //
  }
}